packages/dds/tree/src/test/simple-tree/proxies.spec.ts

```diff
@@ -160,11 +160,96 @@ describe("SharedTreeObject", () => {
+
+	class CustomizedArray extends schemaFactory.array("CustomArray", schemaFactory.number) {
+		public extra = "foo";
+	}
+
+	it("ownKeys", () => {
+		assert.deepEqual(Reflect.ownKeys(hydrate(StructurallyNamedNumberArray, [])), ["length"]);
+		assert.deepEqual(Reflect.ownKeys(hydrate(NumberArray, [])), ["length"]);
+		assert.deepEqual(Reflect.ownKeys(hydrate(CustomizedArray, [])), ["length", "extra"]);
+
+		assert.deepEqual(Reflect.ownKeys(hydrate(StructurallyNamedNumberArray, [5])), [
+			"0",
+			"length",
+		]);
+		assert.deepEqual(Reflect.ownKeys(hydrate(NumberArray, [5])), ["0", "length"]);
+		assert.deepEqual(Reflect.ownKeys(hydrate(CustomizedArray, [5])), ["0", "length", "extra"]);
+	});
+
+	it("in", () => {
+		assert("length" in hydrate(StructurallyNamedNumberArray, []));
+		assert("length" in hydrate(NumberArray, []));
+		assert("length" in hydrate(CustomizedArray, []));
+		assert(!("extra" in hydrate(StructurallyNamedNumberArray, [])));
+		assert(!("extra" in hydrate(NumberArray, [])));
+		assert("extra" in hydrate(CustomizedArray, []));
+		assert(!("extra2" in hydrate(CustomizedArray, [])));
+		assert(!("0" in hydrate(StructurallyNamedNumberArray, [])));
+		assert(!("0" in hydrate(NumberArray, [])));
+		assert(!("0" in hydrate(CustomizedArray, [])));
+
+		assert("0" in hydrate(StructurallyNamedNumberArray, [5]));
+		assert("0" in hydrate(NumberArray, [5]));
+		assert("0" in hydrate(CustomizedArray, [5]));
+	});
+
+	it("length", () => {
+		assert.equal(hydrate(StructurallyNamedNumberArray, []).length, 0);
+		assert.equal(hydrate(StructurallyNamedNumberArray, [1, 2, 3]).length, 3);
+		assert.deepEqual(
+			Reflect.getOwnPropertyDescriptor(hydrate(StructurallyNamedNumberArray, [5]), "length"),
+			Reflect.getOwnPropertyDescriptor([5], "length"),
+		);
+
+		assert.equal(hydrate(NumberArray, []).length, 0);
+		assert.equal(hydrate(NumberArray, [1, 2, 3]).length, 3);
+		assert.deepEqual(
+			Reflect.getOwnPropertyDescriptor(hydrate(NumberArray, [5]), "length"),
+			Reflect.getOwnPropertyDescriptor([5], "length"),
+		);
+
+		// Since getOwnPropertyDescriptor reports length as writable, but its not actually writable, ensure writing it errors:
+		{
+			const array = hydrate(NumberArray, []);
+			assert.throws(() => {
+				// @ts-expect-error length is readonly
+				array.length = 1;
+			});
+		}
+		{
+			const array = hydrate(StructurallyNamedNumberArray, []);
+			assert.throws(() => {
+				// @ts-expect-error length is readonly
+				array.length = 1;
+			});
+		}
+	});
+
+	it("Json stringify", () => {
+		// JSON.stringify uses ownKeys and getOwnPropertyDescriptor
+
+		assert.equal(JSON.stringify(hydrate(StructurallyNamedNumberArray, [])), "[]");
+		assert.equal(JSON.stringify(hydrate(StructurallyNamedNumberArray, [1, 2, 3])), "[1,2,3]");
+		assert.equal(JSON.stringify(hydrate(NumberArray, [])), "{}");
+		assert.equal(JSON.stringify(hydrate(NumberArray, [1, 2, 3])), `{"0":1,"1":2,"2":3}`);
```

Comment:
Ouch.  Classes extend arrays should still stringify as arrays, but we can deal with that separately.

(Now AB#7461)

Reply:
I think thats a more complicated choice than it appears since they can have non numeric own properties that should be included. I have added a comment to your issue.